### DISCUSSION TITLE
Including licenses in wheels for optional dependencies via FetchContent

### DISCUSSION THREAD

**Discussion by minrk** [scikit-build/scikit-build-core#1464] [Q&A]
Summary:

I'd like to make sure I include the licenses of bundled wheel dependencies. The tricky bit is that these are _optional_ dependencies, so the license files aren't available until/unless FetchContent retrieves them. What's the best way to get the license files into wheels when the dependencies are bundled in the wheel?

Detail:

pyzmq optionally bundles two libraries via FetchContent. When it does so, I think the licenses should be added to wheel.license-files metadata. It appears that the license-files globs are [evaluated](https://github.com/scikit-build/scikit-build-core/blob/50a9386015334d56842161225f9e1bb963f1718b/src/scikit_build_core/build/wheel.py#L228-L233) prior to build, making it a bit tricky to make sure these are included when they should be.

What's the best strategy for including license files of optional dependencies that might be fetched during cmake configure? I tried 

```cmake
  file(COPY_FILE "${bundled_libzmq_SOURCE_DIR}/LICENSE" "${CMAKE_CURRENT_SOURCE_DIR}/licenses/LICENSE.zeromq.txt")
```

with

```toml
wheel.license-files = ["licenses/LICENSE*"]
```

but realize the globs are probably evaluated before that COPY_FILE.  Running build a second time (with `--no-isolation`) results in the files being included.

For my official wheels, I realize I can download the licenses by hand in e.g. cibuildwheel before_all, but it would be nice to have a general solution for the standard CMake way of bundling dependencies to also include license files. I don't know if it's as simple as delaying license_file_globs evaluation until after cmake.

**Comment by henryiii** [comment]
Evaluating the globs later would be doable, but it's not really ideal to write the files back into the source folder in the wheel stage. I can think of two options:

The one that works now would be to enable `sdist.cmake`, and then download the source in the SDist step. Then you could even include the relevant parts and be able to build your wheel directly from the SDist without an internet connection.

The other would be a feature I recently realized was missing when looking at hatchling, and have been planning to add. We can expose a `SKBUILD_METADATA_DIR` variable that lets you write files to the metadata directory, and then you'd be able to write to the `licenses` directory instead of something in source. Might require a little though to see how this interacts with metadata 2.2/2.3, which was just added to PyPI and we will likely support generally soon. We add files to the License-File field which is supposed to match between SDist and Wheel unless it's marked (in metadata, not pyproject.toml) as Dynamic. But we can properly document that however it's solved.

**Reply by minrk** [reply]
Writing to metadata during build would be nice. Then the copy-file I have would work with the right destination. I would rather not include the optional sources in the sdists.

**Comment by henryiii** [comment]
I think force-include solves any remaining issues here.
